src/group_theory/double_coset.lean

```diff
@@ -0,0 +1,337 @@
+end
+
+lemma rel_reflex {H K : subgroup G} : reflexive (double_coset_rel H K) :=
+begin
+  rw double_coset_rel,
+  simp only [exists_prop],
+  intro x,
+  use 1,
+  simp only [H.one_mem, K.one_mem, one_mul, exists_eq_right, self_eq_mul_right, and_self],
+end
+
+lemma rel_symm {H K : subgroup G} : symmetric (double_coset_rel H K) :=
+begin
+  rw double_coset_rel,
+  intros x y,
+  simp only [and_imp, exists_prop, forall_exists_index],
+  intros a ha b hb hx,
+  use a⁻¹,
+  have haa := subgroup.inv_mem H ha,
+  simp only [haa, true_and],
+  use b⁻¹,
+  have hbb := subgroup.inv_mem K hb,
+  simp only [hbb, true_and],
+  rw  hx,
+  simp_rw ← mul_assoc,
+  simp only [one_mul, mul_left_inv, mul_inv_cancel_right],
+end
+
+lemma rel_trans {H K : subgroup G} : transitive (double_coset_rel H K) :=
+begin
+  rw double_coset_rel,
+  intros x y z,
+  simp only [and_imp, exists_prop, forall_exists_index],
+  intros a ha b hb hxy c hc d hd hyz,
+  use c * a,
+  have hac := H.mul_mem hc ha,
+  simp only [hac, true_and],
+  use b*d,
+  have hdb :=  (K.mul_mem hb hd),
+  simp only [hdb, true_and],
+  rw  hyz,
+  rw  hxy,
+  simp_rw ← mul_assoc,
+end
+
+lemma rel_is_equiv  {H K : subgroup G} : equivalence (double_coset_rel H K) :=
+⟨rel_reflex, rel_symm, rel_trans⟩
+
+/--The setoid defined by the double_coset relation-/
+def setoid (H K : subgroup G) : setoid G :=
+⟨double_coset_rel H K, rel_is_equiv⟩
+
+/--Quotient of `G` by the double coset relation, i.e. `H \ G / K`-/
+def quotient (H K : subgroup G) : Type* := quotient (setoid H K)
+
+/--The double_coset as an element of `set α` corresponding to `s a t` -/
+def doset (s t : set α) (a : α) : set α := { b : α | ∃ x : (s × t), b = x.1 * a * x.2 }
+
+@[simp]
+lemma doset_mem (s t : set α) (a b : α) :
+  b ∈ (doset s t a) ↔  ∃ x : (s × t), b = x.1 * a * x.2 :=iff.rfl
+
+lemma sub_doset  (H K : subgroup G) (a b : G) (hb : b ∈ doset H.1 K a) :
+  (doset H.1 K b) ⊆ (doset H K a) :=
+begin
+  intro x,
+  simp only [and_imp, exists_prop, forall_exists_index, set_coe.exists, subgroup.mem_carrier,
+  set_like.mem_coe, doset_mem,
+  subgroup.coe_mk] at *,
+  intros g hg,
+  cases hb,
+  rw hb_h at hg,
+  use ⟨g.1 * hb_w.1, hb_w.2 * g.2⟩,
+  simp [hg],
+  simp_rw ← mul_assoc,
+end
+
+--Is this already somewhere? I'd rather not have to prove it
+lemma auxeq : ∀ (a b c d e f : G), a * b * c = d * e * f ↔ b = a⁻¹ * d * e * f * c⁻¹ :=
+begin
+  intros a b c d e f,
+  split; intro h,
+  { assoc_rw h.symm,
+    group,},
+  { rw h,
+    group,},
+end
+
+lemma disjoint_sub (H K : subgroup G) (a b : G) (h : ¬ disjoint (doset H.1 K a ) (doset H K b)) :
+  b ∈ doset H.1 K a :=
+begin
+  rw set.not_disjoint_iff at h,
+  simp only [exists_prop, set_coe.exists, doset_mem, subgroup.mem_carrier, set_like.mem_coe,
+    subgroup.coe_mk] at *,
+  rcases h with ⟨x, ⟨hx, hxx⟩, hk, hhk⟩ ,
+  use ( ⟨hk.1⁻¹ * hx.1, hx.2 * hk.2⁻¹⟩ : H × K),
+  rw hxx at hhk,
+  simp only [subgroup.coe_inv, subgroup.coe_mul],
+  simp_rw  ← mul_assoc,
+  rw ← auxeq,
```

Comment:
I think you can directly use lemmas called something like `inv_mul_eq_iff` and `mul_inv_eq_iff`. (Plus some intermediate `mul_assoc`ing.

Reply:
Yep that worked nicely, thanks. I removed the aux lemma that I was using instead.